minizip fails building on Gentoo because they broke the upstream API for dubious reasons (error: expected initializer before 'OF')

Opening the issue for @Zatsuwar based on https://github.com/godotengine/godot/pull/10428#issuecomment-323783105

----

**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Gentoo, master branch

GCC - gcc (Gentoo 4.9.4 p1.0, pie-0.6.4) 4.9.4
Clan - clang version 3.9.1

Both clang and gcc give-me error.

**Issue description:**
I`m getting errors to compile in my Gentoo, this fallowing errors:

[Initial build] Compiling ==> main/main.cpp In file included from ./thirdparty/minizip/unzip.h:55:0, from ./core/io/file_access_zip.h:38, from main/main.cpp:68: ./thirdparty/minizip/ioapi.h:127:51: error: expected initializer before 'OF' typedef voidpf (ZCALLBACK open_file_func) OF((voidpf opaque, const char filename, int mode)); .....

Complete log: https://ghostbin.com/paste/2kh96

I also tried "minizip=no" but still don't work.

**Steps to reproduce:**
I am running:
```
scons platform=x11 
```
and
```
scons platform=x11 bits=64 tools=no target=release_debug -j 9 use_llvm=yes
```

--- a/thirdparty/minizip/godot-zlib-1.2.4-minizip-unbreak-gentoo.patch
+++ b/thirdparty/minizip/godot-zlib-1.2.4-minizip-unbreak-gentoo.patch
@@ -1,0 +1,27 @@
+diff --git a/thirdparty/minizip/ioapi.h b/thirdparty/minizip/ioapi.h
+index f25ab6464..6043d34ce 100644
+--- a/thirdparty/minizip/ioapi.h
++++ b/thirdparty/minizip/ioapi.h
+@@ -44,6 +44,22 @@
+ #include <stdlib.h>
+ #include "zlib.h"
+ 
++/* GODOT start */
++/* Mighty Gentoo saves the day by breaking the API of their zlib.h,
++ * removing this definition of OF(args) for no practical reason
++ * worth breaking compatibility with all projects that embed minizip
++ * while trying not to diverge too much from upstream zlib.
++ * Cf. https://github.com/godotengine/godot/issues/10539
++ *
++ *   "By and large, this is good open source behaviour, and fits with
++ *    the gentoo _don't fuck with upstream's releases_ philosophy"
++ *                                     -- Gentoo philosopher
++ */
++#ifndef OF /* function prototypes */
++ #define OF(args)  args
++#endif
++/* GODOT end */
++
+ #if defined(USE_FILE32API)
+ #define fopen64 fopen
+ #define ftello64 ftell

--- a/thirdparty/minizip/ioapi.h
+++ b/thirdparty/minizip/ioapi.h
@@ -44,6 +44,22 @@
 #include <stdlib.h>
 #include "zlib.h"
 
+/* GODOT start */
+/* Mighty Gentoo saves the day by breaking the API of their zlib.h,
+ * removing this definition of OF(args) for no practical reason
+ * worth breaking compatibility with all projects that embed minizip
+ * while trying not to diverge too much from upstream zlib.
+ * Cf. https://github.com/godotengine/godot/issues/10539
+ *
+ *   "By and large, this is good open source behaviour, and fits with
+ *    the gentoo _don't fuck with upstream's releases_ philosophy"
+ *                                     -- Gentoo philosopher
+ */
+#ifndef OF /* function prototypes */
+ #define OF(args)  args
+#endif
+/* GODOT end */
+
 #if defined(USE_FILE32API)
 #define fopen64 fopen
 #define ftello64 ftell

